### PR TITLE
Implement Fun Command Cog.

### PR DESCRIPTION
**Pull request by bsrecker**
Create a Fun Command Cog with various commands:

-8ball: A magic eight ball with various funny responses.
-covid: A command to track covid19 related statistics
-ping: Latency of the bot.
-mc: Check the server status for minecraft servers.

### CODE REVIEW

File: src/cogs/fun.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+class Fun(commands.Cog):
+    """Fun Bot Commands"""
+    def __init__(self, bot):
+        self.bot = bot
+
+    @commands.command(name='8ball',
+                      description="Answers a yes/no question.",
+                      brief="Answers from the beyond.",
+                      aliases=['eight_ball', 'eightball', '8-ball'],
+                      pass_context=True)
```
Comment (mudkipdev) [inline src/cogs/fun.py]: context is passed automatically in rewrite

File: src/cogs/fun.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+
+class Fun(commands.Cog):
+    """Fun Bot Commands"""
+    def __init__(self, bot):
+        self.bot = bot
+
+    @commands.command(name='8ball',
+                      description="Answers a yes/no question.",
+                      brief="Answers from the beyond.",
+                      aliases=['eight_ball', 'eightball', '8-ball'],
```
Comment (mudkipdev) [inline src/cogs/fun.py]: What about "8ball" as an alias?

File: src/cogs/fun.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+
+class Fun(commands.Cog):
+    """Fun Bot Commands"""
+    def __init__(self, bot):
+        self.bot = bot
+
+    @commands.command(name='8ball',
+                      description="Answers a yes/no question.",
+                      brief="Answers from the beyond.",
```
Comment (mudkipdev) [inline src/cogs/fun.py]: Do we need both `description` and `brief`?

File: src/cogs/fun.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+import aiohttp
+import discord
+import requests as requests
+from discord.ext import commands
+from datetime import datetime
+
+
+class Fun(commands.Cog):
+    """Fun Bot Commands"""
```
Comment (mudkipdev) [inline src/cogs/fun.py]: Be more descriptive?

File: src/cogs/fun.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                      aliases=['eight_ball', 'eightball', '8-ball'],
+                      pass_context=True)
+    async def eight_ball(self, ctx):
+        possible_responses = [
+            'That is a resounding no',
+            'It is not looking likely',
+            'Too hard to tell',
+            'It is quite possible',
+            'Definitely',
+            'Why don\'t you figure it out yourself you lazy motherfucker',
```
Comment (mudkipdev) [inline src/cogs/fun.py]: Haha a bit too sassy lol? Where'd you get these responses from?

Comment (mudkipdev) [reply src/cogs/fun.py]: https://en.wikipedia.org/wiki/Magic_8-ball#Possible_answers

File: src/cogs/fun.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+            'My sources say no, but they also said Hillary would win.',
+            'Do not swipe right, it\'s your cousin',
+            'Do swipe right, it\'s your cousin!',
+            'Do what Jesus would do, die at the age of 33',
+            'Trump uses me when deciding to go to war'
+
+        ]
+        em = discord.Embed(color=self.bot.config['colors']['default'])
+        em.title = ":mage: The 8 ball wizard says..."
+        em.description = (random.choice(possible_responses) + ", " + ctx.message.author.mention)
```
Comment (mudkipdev) [inline src/cogs/fun.py]: ctx.author.mention is shorter, also we could remove the paranthesis around the variable

File: src/cogs/fun.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        await ctx.send(embed=em)
+
+    @commands.command(name="covid")
```
Comment (mudkipdev) [inline src/cogs/fun.py]: redundant "name="?

File: src/cogs/fun.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    async def covid(self, ctx, *, country=None):
+        """Updated information about Covid-19."""
+        try:
+            if country is None:
+                embed = discord.Embed(title="This command is used like this: ```?covid [country]```", colour=0xff0000,
```
Comment (mudkipdev) [inline src/cogs/fun.py]: Why is "country" optional if we are going to send an error message when it's not included?

File: src/cogs/fun.py [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+                                      timestamp=ctx.message.created_at)
+                await ctx.send(embed=embed)
+
+
+            else:
+                url = f"https://coronavirus-19-api.herokuapp.com/countries/{country}"
+                stats = requests.get(url)
```
Comment (mudkipdev) [inline src/cogs/fun.py]: Requests.get will block the asyncio event loop, use aiohttp instead
https://docs.aiohttp.org/en/stable/

(if you do decide to use aiohttp could you make the client session a property of the bot?)

File: src/cogs/fun.py [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        except:
+            embed3 = discord.Embed(title="Invalid Country Name Try Again..!", colour=0xff0000,
+                                   timestamp=ctx.message.created_at)
+            embed3.set_author(name="Error!")
+            await ctx.send(embed=embed3)
+
+    @commands.command()
+    async def ping(self, ctx):
+        """Pong! Get the bot's response time"""
+        await ctx.channel.purge(limit=1)
```
Comment (mudkipdev) [inline src/cogs/fun.py]: What's this for?

File: src/cogs/fun.py [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        em = discord.Embed(color=discord.Color.green())
+        em.title = ":ping_pong: Pong!"
+        em.description = f'{round(self.bot.latency * 1000)} ms'
+        await ctx.send(embed=em)
+
+    @commands.command()
```
Comment (mudkipdev) [inline src/cogs/fun.py]: Add minecraft alias

File: src/cogs/fun.py [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    async def mc(self, ctx, url=None):
+        """Check the status of your favorite Minecraft Server!"""
+        if url is None:
+            return await ctx.send('Please provide a valid server URL or IP!')
```
Comment (mudkipdev) [inline src/cogs/fun.py]: Again, what's with the non-optional optional stuff?

File: src/cogs/fun.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        else:
+            async with aiohttp.ClientSession() as cs:
+                async with cs.get(f'https://api.mcsrvstat.us/2/{url}') as r:
+                    resp = await r.json()
+
+            if resp["online"] is True:
```
Comment (mudkipdev) [inline src/cogs/fun.py]: if resp["online"]

File: src/cogs/fun.py [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+                    color=discord.Color.blue(),
+                    description=f"Server Online!!! \n Players:{resp['players']['online']}",
+                    timestamp=datetime.utcnow())
+                await ctx.send(embed=embed)
+
+            else:
+                embed = discord.Embed(
+                    title=f"Server status for {url}",
+                    color=discord.Color.blue(),
+                    description="Server is Offline!",
```
Comment (mudkipdev) [inline src/cogs/fun.py]: " and ' are mixed everywhere in this cog, you could do a ctrl + f for them?

**Review by mudkipdev — CHANGES_REQUESTED** [review]
Needs major changes before merging

File: src/cogs/fun.py (hunk 2 above)
Comment (mudkipdev) [reply src/cogs/fun.py]: Nevermind I didn't see the name

**Comment by mudkipdev** [comment]
Should a COVID command be in the fun cog? Personally I think you should save it for another PR

File: src/cogs/fun.py (hunk 5 above)
Comment (bsrecker) [reply src/cogs/fun.py]: Lol I really just added these for some extra fun. Maybe we could add custom responses in the future.

File: src/cogs/fun.py (hunk 10 above)
Comment (bsrecker) [reply src/cogs/fun.py]: This was just to remove the ping command, its not really necessary if we want to leave the commands visible.

File: src/cogs/fun.py [hunk 15]
```diff
@@ -5,64 +5,60 @@
...
-
-        ]
-        em = discord.Embed(color=self.bot.config["colors"]["default"])
-        em.title = ":mage: The 8 ball wizard says..."
-        em.description = (random.choice(possible_responses) + ", " + ctx.author.mention)
+        with open("res/dadjokes.txt").readlines() as jokes:
+            self.jokes = jokes
+
+        with open("res/8ball.txt").readlines() as _8ball_responses:
+            self._8ball_responses = _8ball_responses
```
Comment (bsrecker) [inline src/cogs/fun.py:19]: I think this needs some error handling to cover the case of the files not being accessible.

File: src/cogs/fun.py [hunk 16]
```diff
@@ -5,64 +5,60 @@
...
+
+    @commands.command(aliases = ["joke"])
+    async def dadjoke(self, ctx):
+        embed = discord.Embed(title="Joke", description=random.choice(self.jokes).split("<>"))
```
Comment (bsrecker) [inline src/cogs/fun.py:23]: Isn't self.jokes already a list. What does .split() do here exactly?

File: src/cogs/fun.py [hunk 17]
```diff
@@ -5,64 +5,60 @@
...
+
+    @commands.command(name="8ball", aliases=["eightball"], help="Helps you answer a yes/no question.")
+    async def _8ball(self, ctx):
+        embed = discord.Embed(color=self.bot.config["colors"]["default"])
+        embed.title = ":8ball: The 8 ball says..."
+        embed.description = random.choice(self._8ball_responses) + ", " + ctx.author.mention
         await ctx.send(embed=em)
 
-    @commands.command()
+    @commands.command(help = "Get the bot's response time.")
```
Comment (bsrecker) [inline src/cogs/fun.py:33]: I thought we agreed on using docstrings for help items """Get the bot's response time"""

File: src/cogs/fun.py (hunk 16 above)
Comment (mudkipdev) [reply src/cogs/fun.py:23]: It seperates the joke from the punchline.

**Pull request merged**